keps/sig-auth/0000-20170814-bounding-self-labeling-kubelets.md

```diff
@@ -0,0 +1,163 @@
+
+* The kubelet would transform CSI topology to kubernetes labels before adding them to the `Node` API object.
+* The CSI provisioner would transform CSI topology labels to kubernetes labels when [retrieving topology info from a given node](https://github.com/kubernetes-csi/external-provisioner/blob/v0.5.0-alpha.0/pkg/controller/topology.go#L295-L305).
+* The CSI provisioner would transform kubernetes labels into CSI topology labels when [building required topology from a storage class AllowedTopologies](https://github.com/kubernetes-csi/external-provisioner/blob/v0.5.0-alpha.0/pkg/controller/topology.go#L75-L91).
+* The associated storage class would continue to specify topology in terms of kubernetes label selectors.
+
+Example 1:
+
+* CSI driver returns topology labels `{"acme.com/zone":"A"}`
+* The kubelet records `["acme.com/zone"]` in CSINodeInfo.CSIDriverInfo.TopologyKeys
+* The kubelet transforms the topology info to a `{"acme.com.topology.kubernetes.io/zone":"A"}` label and adds it to the Node object
+* The storage class lists allowed topology selectors of `{"key":"acme.com.topology.kubernetes.io/zone","values":["A"]}`
+* The CSI provisioner transforms the topology key of `"acme.com/zone"` to a label name of `"acme.com.topology.kubernetes.io/zone"` in [getTopologyFromNode](https://github.com/kubernetes-csi/external-provisioner/blob/v0.5.0-alpha.0/pkg/controller/topology.go#L295-L305)
+* The CSI provisioner transforms the storage class selector of `{"key":"acme.com.topology.kubernetes.io/zone","values":["A"]}` to a topology requirement of `{"acme.com/zone":"A"}` when creating the volume
+
+Example 2:
+
+* CSI driver returns topology labels `{"zone":"A"}`
+* The kubelet records `["zone"]` in CSINodeInfo.CSIDriverInfo.TopologyKeys
+* The kubelet transforms the topology info to a `{"topology.kubernetes.io/zone":"A"}` label and adds it to the Node object
+* The storage class lists allowed topology selectors of `{"key":"topology.kubernetes.io/zone","values":["A"]}`
+* The CSI provisioner transforms the topology key of `"zone"` to a label name of `"topology.kubernetes.io/zone"` in [getTopologyFromNode](https://github.com/kubernetes-csi/external-provisioner/blob/v0.5.0-alpha.0/pkg/controller/topology.go#L295-L305)
+
+## Implementation Timeline
+
+v1.13:
+
+* Kubelet deprecates setting labels other than the ones described above via `--node-labels`, and warns when invoked with labels outside that set
+* NodeRestriction admission prevents kubelets from setting labels other than the ones described above on Node *update* (not on Node create)
+
+v1.15:
+
+* Kubelet removes the ability to set labels other than the ones described above via `--node-labels` (deprecation period of 6 months for CLI elements of admin-facing components is complete)
+
+v1.17:
+
+* NodeRestriction admission prevents kubelets from setting labels other than the ones describe above on Node create (oldest supported kubelet running against a v1.17 apiserver is v1.15)
```

Comment:
Same question as above.

Reply:
all labels are "changed" on create from non-existent to present, so all labels on create would be subject to validation